matchzoo/datapack.py

```diff
@@ -0,0 +1,114 @@
+"""Matchzoo DataPack, pair-wise tuple (feature) and context as input."""
+
+import typing
+from pathlib import Path
+
+import dill
+import pandas as pd
+
+
+class DataPack(pd.DataFrame):
+    """
+    Matchzoo DataPack data structure, store dataframe and context.
+
+    Example:
+        >>> # features, context generate by processors.
+        >>> features = [([1,3], [2,3]), ([3,0], [1,6])]
+        >>> context = {'vocab_size': 2000}
+        >>> dp = DataPack(data=features,
+        ...               context=context)
+        >>> dp.context
+        {'vocab_size': 2000}
+        >>> # sample without replacement for generation.
+        >>> type(dp.sample(1))
+        <class 'matchzoo.datapack.DataPack'>
+        >>> dp.size
+        2
+        >>> features, context = dp.unpack()
+    """
+
+    _metadata = ['context']
+
+    DATA_FILENAME = 'data.pkl'
+    CONTEXT_FILENAME = 'context.pkl'
+
+    def __init__(self,
+                 data: list,
+                 context: dict={},
+                 index: list= None,
+                 columns: list=['text_left', 'text_right'],
+                 dtype: object=None,
+                 copy: bool=True):
+        """Initialize."""
+        super(DataPack, self).__init__(data=data,
+                                       index=index,
+                                       columns=columns,
+                                       dtype=dtype,
+                                       copy=copy)
+        if self.shape[1] != 2:
+            raise ValueError(
+                "Pair-wise input expected.")
+        self.context = context
+
+    @property
+    def _constructor(self) -> callable:
+        """Subclass pd.DataFrame."""
+        return DataPack._internal_ctor
+
+    @classmethod
+    def _internal_ctor(cls, *args, **kwargs):
+        """Create subclass inputs to store context."""
+        kwargs['context'] = None
+        return cls(*args, **kwargs)
+
+    @property
+    def size(self) -> int:
```

Comment:
Override `__len__` instead.

Reply:
yes i thought about it, actually `len(DataPack)` is strange from my side, since the term "DataPack" sounds like a multi-dimensional dataframe instead of an array or list. I believe size is better.

Reply:
Using `size` doesn't solve that problem, since your size always returns a single `int`, it doesn't fit the multi-dimensional nature of the structure very well too. Also, `__len__` is used by `__bool__`, so you also can do `if data_pack: ...` this way. However, if you switch to delegation and define a property that exposes the pandas dataframe backend, then you can do `data_pack.df.size` when you need it, but it has different behavior since it multiplies two dimensions, and it means "number of elements" while what you intended to express is "number of samples". Anyways, I think you should take out the inheritance first.